### ISSUE TITLE
[ENH] Annotation of module variables

### ISSUE THREAD

**Issue by jakirkham** [cython/cython#3944]
Have really enjoyed using the pure Python mode annotation support that Cython provides. It has been really great being able to optimize code while still being to engage with other devs in Python.

While most things are already supported in some form or another, there have been a couple things that are missing. One is being able to annotate module variables.

**Is your feature request related to a problem? Please describe.**

In particular I'd like to type `x` and have that type used later in the code. AFAICT Cython currently treats `x` as an `object` instead of a `double`. This also means later code needs to coerce both variables into Python or C types before computing the result.

```cython
x: double = 5.0

def add_y(y: double) -> double:
    return x + y
```

**Describe the solution you'd like**

Ideally annotation like that shown above would result in a variable (like `x`) being typed as a C `double`.

**Describe alternatives you've considered**

Cython has other pure Python acceptable code like `cython.declare`. AFAICT writing code using `declare` does seem to work correctly at the global level.

```cython
from cython import declare

declare(x=double)

x = 5.0

def add_y(y: double) -> double:
    return x + y
```

**Additional context**

NA

**Comment by da-woods** [comment]
I think the reason for not doing so is that it changes `x` from a public attribute of the module to something that's only visible from Cython within the module. For most users that would be quite a dramatic change to happen just because they have an annotation.

With that said, there should possibly be an option to turn it on for people who want to use module annotations.

It might also be possible to make the attribute typed and public with pep562 so that the external interface wouldn't need to change (for people who are using newer Python versions)

**Comment by jakirkham** [comment]
Yeah that makes sense.

I guess with `class` attributes one can manually add `@property` methods or use `declare` (though the latter is an option here).

What if there was a way to include visibility in the annotation like...

```cython
from cython import public

x: public[double]
```

...or similar?

**Comment by da-woods** [comment]
Well at the moment I don't think we can support public module attributes with any type other than `object` - I was just noting that it _might_ be possible in future.

I'd be inclined not to load too much into the annotation - Cython-specific annotations are less useful to anyone else (so might as well just be `declare` or similar)

**Comment by jakirkham** [comment]
Have another question regarding this. Should it be possible to `from ... import x` assuming we've declared it with `visibility="public"`? Or is there more needed? For example should this work?

```python
# mymod.py

from cython import declare, double

x = declare(double, 5.0, visibility="public")
```

```python
In [1]: from mymod import x
```

Right now this is causing an `ImportError`, but am guessing I'm doing something wrong.

**Comment by da-woods** [comment]
In the event that "public" is implemented for module variables then it should be possible to `from mod import var` them.

However there's currently no way to expose a C module variable to Python (and it isn't hugely easy to implement either) so it doesn't work right now

**Comment by jakirkham** [comment]
Yeah that makes sense. Is this still the case if they are actually `PyObject*` like so?

```python
# mymod.py

from cython import declare

d = declare(dict, {"a": 1}, visibility="public")
```

For now have done something sort of hacky like this...

```diff
diff --git a/mymod.py b/mymod.py
index 307fcd7..60bb9c3 100644
--- a/mymod.py
+++ b/mymod.py
@@ -1,3 +1,5 @@
 from cython import declare
+import sys
 
 d = declare(dict, {"a": 1}, visibility="public")
+sys.modules[__name__].d = d
```

This seems to work, but just want to make sure I'm not overlooking a better option.

**Comment by da-woods** [comment]
Yes it's still the case for PyObject. The issue would be that if it was exposed as a public attribute then anyone could assign whatever they like to it (we have no way of controlling this). Therefore the `dict` type would be meaningless. Thus a variable can either be untyped and added to the module dict or typed and private to the module.
